src/CLI/commands/typegen.ts

```diff
@@ -0,0 +1,131 @@
+import { findTsConfigPath } from "CLI/util/findTsConfigPath";
+import { getRojoSourceMap, RojoSourceMap } from "CLI/util/getRojoSourceMap";
+import { getTsConfigProjectOptions } from "CLI/util/getTsConfigProjectOptions";
+import fs from "fs-extra";
+import path from "path";
+import { createProjectData, createProjectProgram, ProjectOptions } from "Project";
+import { createPathTranslator } from "Project/functions/createPathTranslator";
+import { DEFAULT_PROJECT_OPTIONS } from "Shared/constants";
+import ts from "typescript";
+import yargs from "yargs";
+
+const BANNED_SERVICES = new Set([
+	// StarterPlayer contents are cloned elsewhere at runtime. They should not be accessed directly.
+	"StarterPlayer",
+]);
+
+function buildSignatures(children: ReadonlyArray<RojoSourceMap>, path: ReadonlyArray<string>) {
+	const seen = new Set<string>();
+	const members = new Array<ts.PropertySignature>();
+	for (const child of children) {
+		if (child.name === "node_modules") continue;
+		if (seen.has(child.name)) continue;
+		seen.add(child.name);
+		members.push(buildPropertySignature(child, path.concat(child.name)));
+	}
+	return members;
+}
+
+function buildTypeNode(sourceMap: RojoSourceMap, path: ReadonlyArray<string>) {
+	const types = new Array<ts.TypeNode>();
+	types.push(ts.factory.createTypeReferenceNode(sourceMap.className, undefined));
+	if (sourceMap.children) {
+		types.push(ts.factory.createTypeLiteralNode(buildSignatures(sourceMap.children, path)));
+	}
+	return ts.factory.createIntersectionTypeNode(types);
+}
+
+function buildPropertySignature(sourceMap: RojoSourceMap, path: ReadonlyArray<string>) {
+	// Workspace.Terrain is `readonly` and TypeScript will complain that all declarations must have identical modifiers
+	const modifiers = new Array<ts.Modifier>();
+	if (path[0] === "Workspace" && path[1] === "Terrain") {
+		modifiers.push(ts.factory.createModifier(ts.SyntaxKind.ReadonlyKeyword));
+	}
+
+	return ts.factory.createPropertySignature(
+		modifiers,
+		ts.isIdentifierText(sourceMap.name, undefined) ? sourceMap.name : `"${sourceMap.name}"`,
+		undefined,
+		buildTypeNode(sourceMap, path),
+	);
+}
+
+interface TypeGenFlags {
+	project: string;
+	rojo?: string;
+}
+
+/**
+ * Defines the behavior for the `rbxtsc typegen` command.
+ */
+export = ts.identity<yargs.CommandModule<object, TypeGenFlags>>({
+	command: ["typegen"],
+
+	describe: "Creates a `src/services.d.ts` file from `rojo sourcemap`",
```

Comment:
Maybe this should be configurable? I have my globals at `src/shared/globals.d.ts`